idseq_pipeline/commands/non_host_alignment_functions.py

```diff
@@ -709,28 +633,83 @@ def run_gsnapl_chunk(part_suffix, remote_home_dir, remote_index_dir, remote_work
             write_to_log("starting alignment for chunk %s on machine %s" % (chunk_id, gsnapl_instance_ip))
             execute_command(remote_command(commands, key_path, remote_username, gsnapl_instance_ip))
             # check if every row has correct number of columns (12) in the output file on the remote machine
-            verification_command = "awk '{print NF}' %s | sort -nu | head -n 1" % remote_outfile
+            verification_command = "awk '{print NF}' %s | sort -nu | head -n 1" % multihit_remote_outfile
             min_column_number_string = execute_command_with_output(remote_command(verification_command, key_path, remote_username, gsnapl_instance_ip))
             min_column_number = interpret_min_column_number_string(min_column_number_string, correct_number_of_output_columns, try_number)
             try_number += 1
-        # move output from remote machine to s3
+        # move output from remote machine to local
         assert min_column_number == correct_number_of_output_columns, "Chunk %s output corrupt; not copying to S3. Re-start pipeline to try again." % chunk_id
         with iostream:
-            execute_command(scp(key_path, remote_username, gsnapl_instance_ip, remote_outfile, local_outfile))
-        # Deduplicate m8 input. Sometimes GSNAPL outputs multiple consecutive lines for same original read and same accession id. Count functions expect only 1 (top hit).
-        execute_command("echo '' >> %s;" % local_outfile) # add a blank line at the end of the file so S3 copy doesn't fail if output is empty
-        deduplicate_m8(os.path.join(CHUNKS_RESULT_DIR, outfile_basename), os.path.join(CHUNKS_RESULT_DIR, dedup_outfile_basename))
-        with iostream:
-            execute_command("aws s3 cp --quiet %s/%s %s/" % (CHUNKS_RESULT_DIR, dedup_outfile_basename, SAMPLE_S3_OUTPUT_CHUNKS_PATH))
-    with iostream:
-        execute_command("sed -i '$ {/^$/d;}' %s" % os.path.join(CHUNKS_RESULT_DIR, dedup_outfile_basename)) # remove blank line from end of file
+            execute_command(scp(key_path, remote_username, gsnapl_instance_ip, multihit_remote_outfile, multihit_local_outfile))
+
+        # Deduplicate multihit m8 by using taxonomy info
+        call_hits_m8(multihit_local_outfile, dedup_multihit_local_outfile, multihit_summary_file)
+
+        # copy outputs to S3
+        for f in [multihit_local_outfile, dedup_multihit_local_outfile, multihit_summary_file]:
+            with iostream:
+                execute_command("aws s3 cp --quiet %s %s/" % (f, SAMPLE_S3_OUTPUT_CHUNKS_PATH))
+
     write_to_log("finished alignment for chunk %s" % chunk_id)
-    return os.path.join(CHUNKS_RESULT_DIR, dedup_outfile_basename)
+    return multihit_local_outfile
 
 
+def call_hits_m8(input_m8, output_m8, output_summary):
+    # Get lineage info
+    lineage_path = fetch_reference(LINEAGE_SHELF)
+    lineage_map = shelve.open(lineage_path)
```

Comment:
This is running concurrently in 64 threads.    Opening the same file from 64 threads doesn't seem good.   But, what does work, according to shelve docs, and based on our experience in accessionid2seq_functions.py, is opening this in the function that spawns all these threads, and passing the dict down as a param all the way through to here.

Reply:
fixed